### PR TITLE
Feature, worker EBS volume encryption

### PR DESCRIPTION
**Pull request by adamjohnson01**
## Description

Fix for issue #952 

I have added a new option called **volumeKmsKeyID** which can be used with the existing **volumeEncrypted** option to encrypt the worker EBS volumes using a Customer Managed Key from KMS.

### Checklist
- [x] Code compiles correctly (i.e `make build`)
- [x] Added tests that cover your change (if possible)
- [x] All unit tests passing (i.e. `make test`)
- [x] Added/modified documentation as required (such as the `README.md`, and `examples` directory)
- [x] Manually tested

### CODE REVIEW

**Comment by errordeveloper** [comment]
@adamjohnson01 could you please describe how did you test this manually? We might want to capture this to add an integration test case in the future.

**Comment by adamjohnson01** [comment]
> @adamjohnson01 could you please describe how did you test this manually? We might want to capture this to add an integration test case in the future.

@errordeveloper, sure no problem.

Default Encryption
- create a nodegroup with just volumeEncrypt: true
- wait for worker node
- get volume-id from worker instance using aws ec2 describe-instances
- describe volume info using aws ec2 describe-volumes
- describe default kms key using aws kms describe-key --key-id alias/aws/ebs
- check that "Encrypted" is true and that KmsKeyId matches the default key ID

CMK Encryption
- create a nodegroup with volumeEncrypt: true and volumeKmsKeyID: KEY_ID
- wait for worker node
- get volume-id from worker instance using aws ec2 describe-instances
- describe volume info using aws ec2 describe-volumes
- describe default kms key using aws kms describe-key --key-id alias/CMK_ALIAS
- check that "Encrypted" is true and that KmsKeyId matches the CMK ID

File: pkg/ami/api.go [hunk 1]
```diff
@@ -68,6 +68,11 @@ func Use(ec2api ec2iface.EC2API, ng *api.NodeGroup) error {
 		}
 		if ng.VolumeEncrypted == nil {
 			ng.VolumeEncrypted = output.Images[0].BlockDeviceMappings[0].Ebs.Encrypted
+		} else {
+			// VolumeEncrypted cannot be false if the AMI being used is already encrypted.
+			if api.IsDisabled(ng.VolumeEncrypted) && api.IsEnabled(output.Images[0].BlockDeviceMappings[0].Ebs.Encrypted) {
```
Comment (cPu1) [inline pkg/ami/api.go]: Minor nit: `output.Images[0].BlockDeviceMappings[0].Ebs.Encrypted` is used in the corresponding `if` clause, so it'd help to assign it to a variable.

Comment (adamjohnson01) [reply pkg/ami/api.go]: @cPu1 do you mean something like this?

```
amiEncrypted := output.Images[0].BlockDeviceMappings[0].Ebs.Encrypted
if ng.VolumeEncrypted == nil {
	ng.VolumeEncrypted = amiEncrypted
} else {
	// VolumeEncrypted cannot be false if the AMI being used is already encrypted.
	if api.IsDisabled(ng.VolumeEncrypted) && api.IsEnabled(amiEncrypted) {
		return fmt.Errorf("%q is an encrypted AMI and volumeEncrypted has been set to false", ng.AMI)
	}
}
```

Comment (cPu1) [reply pkg/ami/api.go]: @adamjohnson01 yes.

File: pkg/cfn/builder/nodegroup.go [hunk 2]
```diff
@@ -114,14 +114,28 @@ func (n *NodeGroupResourceSet) addResourcesForNodeGroup() error {
...
 	if volumeSize := n.spec.VolumeSize; volumeSize != nil && *volumeSize > 0 {
-		launchTemplateData.BlockDeviceMappings = []gfn.AWSEC2LaunchTemplate_BlockDeviceMapping{{
-			DeviceName: gfn.NewString(*n.spec.VolumeName),
-			Ebs: &gfn.AWSEC2LaunchTemplate_Ebs{
-				VolumeSize: gfn.NewInteger(*volumeSize),
-				VolumeType: gfn.NewString(*n.spec.VolumeType),
-				Encrypted: gfn.NewBoolean(*n.spec.VolumeEncrypted),
-			},
-		}}
+		if api.IsSetAndNonEmptyString(n.spec.VolumeKmsKeyID) {
```
Comment (martina-if) [inline pkg/cfn/builder/nodegroup.go]: You can do here something like:

```
launchTemplateData.BlockDeviceMappings = []gfn.AWSEC2LaunchTemplate_BlockDeviceMapping{{
				DeviceName: gfn.NewString(*n.spec.VolumeName),
				Ebs: &gfn.AWSEC2LaunchTemplate_Ebs{
					VolumeSize: gfn.NewInteger(*volumeSize),
					VolumeType: gfn.NewString(*n.spec.VolumeType),
					Encrypted: gfn.NewBoolean(*n.spec.VolumeEncrypted),
				},
			}}
if api.IsSetAndNonEmptyString(n.spec.VolumeKmsKeyID) {
  launchTemplateData.BlockDeviceMappings.Ebs.KmsKeyId = gfn.NewString(*n.spec.VolumeKmsKeyID)
}
```

Comment (adamjohnson01) [reply pkg/cfn/builder/nodegroup.go]: Ah cool, did not realise I could do it that way. Thanks!

**Comment by adamjohnson01** [comment]
@martina-if, I have added tests and updated the launchTemplateData.BlockDeviceMappings section.

File: pkg/cfn/builder/nodegroup.go [hunk 3]
```diff
@@ -114,28 +114,18 @@ func (n *NodeGroupResourceSet) addResourcesForNodeGroup() error {
 	}
 
 	if volumeSize := n.spec.VolumeSize; volumeSize != nil && *volumeSize > 0 {
+		launchTemplateData.BlockDeviceMappings = []gfn.AWSEC2LaunchTemplate_BlockDeviceMapping{{
+			DeviceName: gfn.NewString(*n.spec.VolumeName),
+			Ebs: &gfn.AWSEC2LaunchTemplate_Ebs{
+				VolumeSize: gfn.NewInteger(*volumeSize),
+				VolumeType: gfn.NewString(*n.spec.VolumeType),
+				Encrypted: gfn.NewBoolean(*n.spec.VolumeEncrypted),
+				KmsKeyId: gfn.NewString(*n.spec.VolumeKmsKeyID),
```
Comment (martina-if) [inline pkg/cfn/builder/nodegroup.go]: I think you forgot to remove the kmskeyid from the original object

Comment (adamjohnson01) [reply pkg/cfn/builder/nodegroup.go]: Ah, thanks! I have updated.

**Pull request merged**